cql3/expr/expression.cc

```diff
@@ -171,7 +171,16 @@ static managed_bytes_opt get_value(const column_maybe_subscripted& col, const co
             const auto deserialized = cdef->type->deserialize(managed_bytes_view(*data.other_columns[index]));
             const auto& data_map = value_cast<map_type_impl::native_type>(deserialized);
             const auto key = evaluate(s->sub, options);
+            // When s->sub is a constant, we should have already verified
+            // earlier that it has the right type (key.type == col_type->
+            // name_comparator()) and not a null or unset value. If we
+            // find an inappropriate key here, we'll assume it was dynamically
+            // calculated, so this is not an invalid request - it's simply
+            // a non-matching item and we should return std::nullopt.
             auto&& key_type = col_type->name_comparator();
+            if (key.is_null_or_unset() || key_type != key.type) {
```

Comment:
The test for key_type should be during prepare, the types are known at that time.

Reply:
It is checked at prepare time for constants, but I guess somebody can still try to bind a string to a bind marker that's supposed to hold an int? And we currently only validate that at evaluation stage

Reply:
You're right. We actually already have those type tests there, although it's hard to confirm for prepared statements because I can't get the driver to send the wrong type for a bound variable. The test here was more of a sanity check - not not do invalid things (and potentially crash) if we ever get this wrong. But maybe as you said for UNSET_VALUE, we can add this as a sanity check and throw a logic_error in this case - and not silently ignore it.

I could do an assert() that the type is correct, but then it's a guaranteed crash, which isn't that much better than a potential crash :-(

Reply:
> It is checked at prepare time for constants, but I guess somebody can still try to bind a string to a bind marker that's supposed to hold an int? And we currently only validate that at evaluation stage

We don't really know it's a string, just that it's an int with sizeof != 4. The binary protocol is untyped apart from the metadata returned by PREPARE.

Reply:
(so it's not a type error, just a serialization error).